### PR TITLE
Cloud functions and rules

### LINKED ISSUE #69: Figure out how to update study metadata firestore

### LINKED ISSUE #84: adjust users and studies schemas to support cross-browser and access control
Currently, we have two collections in Firestore: `users` and `studies`.

`studies` is a collection of documents indexed by an add-on ID:

```
{
  addonId: "rally-test-extension@rally.mozilla.org",
  authors: {
    name: "Rally Developers"
  },
  dataCollectionDetails: [ "test data" ],
  description: "Test",
  downloadLink: "http://example.com",
  endDate: "2023-01-01",
  icons: {
    32: "http://example.com/32.png",
    64: "http://example.com/32.png",
    128: "http://example.com/32.png",
  },
  name: "Rally Test",
  studyDetailsLink: "https://example.com",
  studyEnded: false,
  studyPaused: false,
  tags: [ "testing" ],
  version: "0.0.1"
}
```

`users` is a collection of documents indexed by an internal Firebase UID:

```
{
  createdOn: "2021-01-01",
  demographicsData: {
    age: "25_34",
    gender: "neither",
    race: [ "american_indian_or_alaska_native" ],
    school: "some_college_but_no_degree_or_in_progress",
    zipcode: "90210"
  },
  enrolled: true,
  enrolledStudies: {
    "rally-test-extension@rally.mozilla.org":
      enrolled: true,
      joinedOn: "2021-01-01",
  },
  onboarded: true,
  uid: "..."
}
```

This works fine in a world where extensions can only have a single addon ID, however this will no longer be the case when we have Chrome support. To support this, I suggest that we make the `studies` key a neutral identifier, remove the `addonId` and `downloadLink` keys, and add an `extensionIds` and `downloadLinks` map:

```
{
  extensionIds: {
    firefox: "rally-test-extension@rally.mozilla.org",
    chrome: "nffpaknbpcignmifkebibbncobabkdji",
  },
  authors: {
    name: "Rally Developers"
  },
  dataCollectionDetails: [ "test data" ],
  description: "Test",
  downloadLinks: {
    firefox: "https://addons.mozilla.org/...",
    chrome: "https://chrome.google.com/webstore/...",
  },
  endDate: "2023-01-01",
  icons: {
    32: "http://example.com/32.png",
    64: "http://example.com/32.png",
    128: "http://example.com/32.png",
  },
  name: "Rally Test",
  studyDetailsLink: "https://example.com",
  studyEnded: false,
  studyPaused: false,
  tags: [ "testing" ],
  version: "0.0.1"
}
```

For the `users` document, we should remove the `enrolledStudies` key entirely and make it a reference to a subcollection instead; this way we can share just the study consent details with study extensions, and not allow them access to user info. I suggest that we rename it something like `userStudies`, since `enrolledStudies` implies they are all enrolled when in fact they each have individual `enrolled` keys. This collection is really a record of every study that the user has ever enrolled in, but not necessarily a record of every currently-enrolled study.

Finally, we need some way to tie together Rally IDs with the Firebase UID. These should not be exposed to the normal user account, only the extensions need to see this. This could be another subcollection on the `users` document, or it could be a totally separate collection entirely. I am leaning towards having a `rallyIds` reference on the `user` document which points to a subcollection containing the list of Rally IDs per-study.

Each user document would look something like this:

```
{
  createdOn: "2021-01-01",
  demographicsData: {
    age: "25_34",
    gender: "neither",
    race: [ "american_indian_or_alaska_native" ],
    school: "some_college_but_no_degree_or_in_progress",
    zipcode: "90210"
  },
  enrolled: true,
  onboarded: true,
  rallyIds: <reference>,
  uid: "...",
  userStudies: <reference>,
}
```

`rallyIds` would be indexed by the neutral study name used by the `studies` collection, and contain a document with details on this Rally ID:

```
{
  studyName: "...",
  createdOn: "2021-01-01"
}
```

`userStudies` would also be indexed by the neutral study name used by the `studies` collection, and contain the equivalent of `enrolledStudies` in the original `user` document:

```
{
  enrolled: true,
  joinedOn: "2021-01-01",
}
```

The reason subcollections are used here is because the access level of user accounts vs. study extension accounts are overlapping but distinct:

1. users need access to everything *except* rally IDs
2. study extensions need access to rally IDs and their own study document, but *nothing else*

The advantage to using subcollections is that we can have a single place that users can change enrolled status, and study extensions can watch that document for enrollment changes. Study extensions also need to be able to add and remove Rally IDs. Neither of them need to be able to cross-reference Rally IDs with user identities; that is something we do administratively on the backend in order to deliver givebacks.

Additionally, Rally needs to be able to identify Rally IDs associated with a user to process deletion requests.

The above can be enforced with Firebase Security Rules (which we can run tests against using the emulators).

### PR DESCRIPTION
**Pull request by rhelmer**
This depends on https://github.com/mozilla-rally/rally-web-platform/pull/100, addresses #55, closes #84, and closes #69.

This implements everything from the cloud function proposal required to create new users:

- [x] `addRallyUserToFirestore` cloud  function  - Auth `onCreate` to create all firestore docs
- [x] new user studies subcollection, removing old enrolledStudies map in user doc
- [x] move svelte app from enrolledStudies to user studies subcollection
- [x] update rally-sdk in test extensions to use user studies subcollection
- [x] add cloud firestore rules to enforce the above 

This also adds an `npm run watch` command which runs the Firebase emulators and watches for changes to rules, cloud functions, static data (studies), and the website, which makes it much easier to make+test changes like this.

### CODE REVIEW

**Comment by rhelmer** [comment]
OK so the test is a lot more reliable overall now, still hitting odd situations where this notification is obscuring the Leave button, somehow:

image

**Comment by rhelmer** [comment]
This seems to work well with the latest rally-sdk (https://github.com/mozilla-rally/rally-sdk/commit/d16a00014e4fe84fc0e2a8feb66043da59e0c5e8 as of this writing), the tests seem more reliable so far.

We still need to implement Glean pings and deletion requests, this one is already getting quite large so I'd like to do that as a separate bit of work.

**Comment by rhelmer** [comment]
> We still need to implement Glean pings and deletion requests, this one is already getting quite large so I'd like to do that as a separate bit of work.

Filed https://github.com/mozilla-rally/rally-web-platform/issues/108 to track that.

File: package.json [hunk 1]
```diff
@@ -4,9 +4,13 @@
   "scripts": {
     "pre-build": "bash scripts/reset-build-directories.sh && bash scripts/add-protocol.sh && node scripts/add-version-object.js",
     "dev": "npm run pre-build && svelte-kit dev",
+    "watch": "firebase emulators:exec --ui 'npm run watch:functions & npm run watch:web'",
     "build": "npm run pre-build && svelte-kit build",
     "build:functions": "cd functions && npm install && npm run build && cd ..",
-    "build:test:integration": "npm run pre-build && svelte-kit build -- --config-integration-test-mode",
+    "build:web:emulator": "npm run pre-build && svelte-kit build -- --config-integration-test-mode",
+    "watch:functions": "watch 'npm run build:functions && sleep 1 && npm run load:data' ./functions/src",
+    "watch:web": "watch 'svelte-kit dev -- --config-integration-test-mode' ./src",
```
Comment (hamilton) [inline package.json:12]: Should we deprecate `dev` here / everywhere in favor of `watch`, or vice versa?

**Review by hamilton — APPROVED** [review]
This is great – so many improvements! Had a few comments but looks good to me

**Review by Carla-Moz — APPROVED** [review]
Working for me and the README looks greats. Thanks!

**Comment by rhelmer** [comment]
OK I've addressed review comments, thanks! I buttoned up the last few things we discussed in Slack, and  I removed the single-study hardcoding and it should in theory support multiple studies (we have tests with a single study, I haven't tried multiple yet).

This wfm with `npm run dev` under the VSCode debugger, and tests are passing on Circle now. The only thing I am going to defer is to not generate tokens until the SDK specifically asks (this requires an SDK change, https://github.com/mozilla-rally/rally-sdk/issues/13)

**Pull request merged**